qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py

```diff
@@ -150,30 +124,58 @@ def run(self, dag):
-            logger.info("Skipping pass because no basis is set")
+        if self._basis_gates is None and self._target is None:
+            logger.info("Skipping pass because no basis or target is set")
             return dag
 
         runs = dag.collect_1q_runs()
         for run in runs:
-            # SPECIAL CASE: Don't bother to optimize single U3 gates which are in the basis set.
-            #     The U3 decomposer is only going to emit a sequence of length 1 anyhow.
-            if "u3" in self._target_basis and len(run) == 1 and isinstance(run[0].op, U3Gate):
-                # Toss U3 gates equivalent to the identity; there we get off easy.
-                if np.allclose(run[0].op.to_matrix(), np.eye(2), 1e-15, 0):
-                    dag.remove_op_node(run[0])
-                    continue
-                # We might rewrite into lower `u`s if they're available.
-                if "u2" not in self._target_basis and "u1" not in self._target_basis:
-                    continue
-
-            new_basis, new_circ = self._resynthesize_run(run)
-
-            if new_circ is not None and self._substitution_checks(dag, run, new_circ, new_basis):
+            qubit = dag.qubits.index(run[0].qargs[0])
+            new_circ = self._resynthesize_run(run, qubit)
+
+            if self._target is None:
+                basis = self._basis_gates
+            else:
+                basis = self._target.operation_names_for_qargs((qubit,))
+
+            if new_circ is not None and self._substitution_checks(dag, run, new_circ, basis, qubit):
                 new_dag = circuit_to_dag(new_circ)
                 dag.substitute_node_with_dag(run[0], new_dag)
                 # Delete the other nodes in the run
                 for current_node in run[1:]:
                     dag.remove_op_node(current_node)
 
         return dag
+
+
+def _possible_decomposers(basis_set):
+    decomposers = []
+    euler_basis_gates = one_qubit_decompose.ONE_QUBIT_EULER_BASIS_GATES
+    for euler_basis_name, gates in euler_basis_gates.items():
+        if set(gates).issubset(basis_set):
+            decomposer = one_qubit_decompose.OneQubitEulerDecomposer(euler_basis_name)
+            decomposers.append(decomposer)
+    return decomposers
+
+
+def _error(circuit, target, qubit):
+    """
+    Calculate a rough error for a `circuit` that runs on a specific
+    `qubit` of `target` (circuit could also be a list of DAGNodes)
+
+    Use basis errors from target if available, otherwise use length
+    of circuit as a weak proxy for error.
+    """
+    if target is None:
+        return len(circuit)
+    else:
+        if isinstance(circuit, list):
+            gate_errors = [
+                1 - getattr(target[node.name].get((qubit,)), "error", 0.0) for node in circuit
+            ]
+        else:
+            gate_errors = [
+                1 - getattr(target[inst.operation.name].get((qubit,)), "error", 0.0)
+                for inst in circuit
+            ]
```

Comment:
I'm wondering if we want to fall back to using a length heuristic here if the error rates are 0? Like I'm wondering what we want to do if the target has no error rates set and the local basis is defined as `["rz"," sx", "x", "u"]` The error score for all the matching decompositions will be equal and the output will be based on insertion order in the target (or at least I think it will).

Reply:
good point, I'll update this and make a test

Reply:
updated in https://github.com/Qiskit/qiskit-terra/pull/8917/commits/4322296ac25809872c62e4c9e6dbc28252801713